src/TestFramework/TestFramework.Extensions.Csv/CsvDataSourceAttribute.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Data;
+using System.Data.OleDb;
+using System.Globalization;
+using System.Reflection;
+
+namespace Microsoft.VisualStudio.TestTools.UnitTesting;
+
+/// <summary>
+/// Attribute to define dynamic data from a CSV file for a test method.
+/// </summary>
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = false/*TODO: Do we want to allow multiple*/)]
+public sealed class CsvDataSourceAttribute : Attribute, ITestDataSource
+{
+    // Template used to map from a filename to a DB connection string
+    private const string CsvConnectionTemplate = "Provider=Microsoft.Jet.OLEDB.4.0;Data Source={0};Persist Security Info=False;Extended Properties=\"text;HDR=YES;FMT=Delimited\"";
+    private const string CsvConnectionTemplate64 = "Provider=Microsoft.Ace.OLEDB.12.0;Data Source={0};Persist Security Info=False;Extended Properties=\"text;HDR=YES;FMT=Delimited\"";
+
+    public CsvDataSourceAttribute(string fileName)
+        => FileName = fileName;
+
+    internal string FileName { get; }
+
+    IEnumerable<object?[]> ITestDataSource.GetData(MethodInfo methodInfo)
+    {
+        // TODO: Avoid using OleDb and instead parse Csv directly. Maybe use https://www.nuget.org/packages/CsvHelper? Write our own?
+        // When that happens, the OleDb dependency in the csproj should be
+        // removed, and relevant tests for CsvDataSource should run on Linux.
+
+        // We better work with a full path, if nothing else, errors become easier to report
+        string fullPath = Path.GetFullPath(FileName);
+        if (!File.Exists(fullPath))
+        {
+            // TODO: Localize.
+            throw new FileNotFoundException($"Csv file '{fullPath}' cannot be found.", fullPath);
+        }
+
+        string tableName = Path.GetFileName(fullPath).Replace('.', '#');
+
+        // We can map simplified CSVs to an OLEDB/Text connection, then proceed as normal
+        using OleDbConnection connection = new();
+        using OleDbDataAdapter dataAdapter = new();
+        using OleDbCommandBuilder commandBuilder = new();
+        using OleDbCommand command = new();
```

Comment:
Nit: I usually prefer to initialize variable close to where they are used rather than having init blocks and usage blocks.

Reply:
Agreed. This was almost a copy/paste from the existing code.